### DISCUSSION TITLE
Update the gsi automatically

### DISCUSSION THREAD

**Discussion by clemzangrilli** [sensedeep/dynamodb-onetable#403] [Q&A] [answered]
Hello everyone ! 

I have a gsi which is depending on an attribute. When I'm updating this attribute, the gsi is not updated automatically. How do you do that ? 

# More explications below

### I have the model User with one gsi : 

```
User: {
      pk: { type: String, value: 'user#${orgId}' },
      sk: { value: 'user#${email}', type: String },

      gs1pk: { type: String, value: 'user#${orgId}#${teamId}' },
      gs1sk: {
        value: 'user#${firstName}',
        type: String,
      },

      id: { type: String },
      orgId: { type: String, required: true },
      teamId: { type: String },
      email: { type: String, required: true },
}
```

### If I update the user with a new teamId, the gs1pk is not updated : 

```
const userUpdated = await USER_TABLE.update({orgId: "id", email: "email"}, {
    set: {
      teamId: "newId"
    }
});
```

### The only way to update the gsi, is to provide the attributes of the gsi in the properties : 
```
const userUpdated = await USER_TABLE.update({orgId: "id", email: "email", teamId: "newId"}, {
    set: {
      teamId: "newId"
    }
});
```
I understand that the old team id is needed to dynamodb, to find the gsi to update. But if we provide the oldId in properties, the gsi is still not updated. We must provide the new teamId in properties to have the gsi updated.

The next problem is : if I add a gsi to the user table in the future, I would have to change all my updates call, to include a new param. 

**Is there an easy way to update the gsi automatically ?**

Thank you for your help !

**Comment by TheSmartMonkey** [comment]
+1

**Comment by kharchafiabdallah** [comment]
+1

**Comment by mobsense** [comment, marked as answer]
When you are using value templates, you MUST provide the values for those via the properties. Set expressions are intended for more complex expressions where you want DynamoDB to perform an expression to determine the value.

Think of the GSI as just an index. You have a primary and a secondary index.

When you update the item by providing PK & SK, you identify the item. When you provide a new value for the team, the item is updated and a new GSI index entry is created. 

So all you need to do is:

    USER_TABLE.update({orgId: "id", email: "email", teamId: "newId"})

No set required:

**Reply by clemzangrilli** [reply]
Ok thank you for your response ! I understand that we need to put all variables that compose the indexes (primary key, sort key, GSI, ...) in the properties. 

You say that set are not mandatory and are "intended for more complex expressions". 

What is the best practice ?  

```
const userUpdated = await USER_TABLE.update({orgId: "id", email: "email, teamId: "newId"}, {
    set: {
      basicAttribute: "value"
    }
});
```

OR

```
const userUpdated = await USER_TABLE.update({orgId: "id", email: "email, teamId: "newId", basicAttribute: "value"});
```

**Reply by mobsense** [reply]
The latter. It is simpler to read.

Update is designed to update all the supplied attributes. If you set an attribute to null, it is removed (if table params nulls is false).
It will extract the keys from the given properties and update the others.

**Reply by clemzangrilli** [reply]
Good to know, thank you !
